external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp

```diff
@@ -513,54 +554,159 @@ class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
+              // userOp is acceptable.  Keep it as a trailingOp, put it on the
               // worklist.  Add its operands to inputNodes unless they come
-              // from other secondOps (indicated by being in resultNodes).
-              // If all the users of any resultNode are in secondOps, there's
+              // from other trailingOps (indicated by being in resultNodes).
+              // If all the users of any resultNode are in trailingOps, there's
               // no need to return it so remove from resultNodes.  Finally,
               // add all userOp's results to resultNodes.
               if (!skip) {
-                secondOps.insert(userOp);
+                if (isa<tosa::TransposeOp>(userOp)) {
+                  specialCaseForTranspose(userOp, trailingOps);
+                }
+                // General case.
+                trailingOps.insert(userOp);
                 worklist.push_back(userOp);
-                for (Value opnd : userOp->getOperands()) {
-                  if (!resultNodes.contains(opnd)) {
+                for (Value opnd : userOp->getOperands())
+                  if (!resultNodes.contains(opnd))
                     inputNodes.insert(opnd);
-                  }
-                }
-                for (const Value &val : resultNodes) {
+                for (const Value &val : resultNodes)
                   if (llvm::all_of(val.getUsers(), [&](Operation *u) {
-                        return secondOps.contains(u);
-                      })) {
+                        return trailingOps.contains(u);
+                      }))
                     resultNodes.remove(val);
-                  }
-                }
-                for (auto res : userOp->getResults()) {
+                for (auto res : userOp->getResults())
                   resultNodes.insert(res);
-                }
               }
             }
           }
         }
 
         // Make the outlined function from the ops we've gathered.
-        outlineConvPartOps(convOp, secondOps.getArrayRef(), frontOps,
-                           inputNodes.getArrayRef(), resultNodes.getArrayRef(),
+        outlineConvPartOps(anchorOp, trailingOps.getArrayRef(),
+                           leadingOps.getArrayRef(), inputNodes.getArrayRef(),
+                           resultNodes.getArrayRef(),
                            std::string(func.getSymName()) + strCount,
-                           candidates);
+                           attributeName(), candidates);
         // Outlining will erase nodes and thus perturb the walk, so
         // signal interrupted to exit it and restart.
         return WalkResult::interrupt();
       };
 
-      // Walk until we've outlined all the Conv2D ops we can.
+      // Walk until we've outlined all the anchor ops we can.
       while (func.walk(callback).wasInterrupted()) {
       }
     }
   }
 };
 
+class TosaPartitionPassWithOptions
```

Comment:
The variant of TosaPartitionPass that obeys the command-line sub-options.  It's a subclass of TosaPartitionPass and overrides the key predicates.  The no-args constructor uses the option member variables as they are, and there's another constructor to pass them in.